enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,401 @@
+approvers:
+  - @dhellmann
+api-approvers:
+  - TBD
+creation-date: 2024-01-16
+last-updated: 2024-01-23
+tracking-link:
+  - https://issues.redhat.com/browse/OCPSTRAT-473
+# see-also:
+#   - "/enhancements/this-other-neat-thing.md"
+# replaces:
+#   - "/enhancements/that-less-than-great-idea.md"
+# superseded-by:
+#   - "/enhancements/our-past-effort.md"
+---
+
+# Multus CNI for MicroShift
+
+## Summary
+
+Currently MicroShift ships [ovn-kubernetes](https://github.com/openshift/ovn-kubernetes) (ovn-k)
+CNI responsible for connectivity within and outside the cluster.
+There are users that have needs beyond what ovn-k offers like adding more interfaces to the Pods.
+Some example requirements are connecting Pod to host's bridge interface or setting up complex networking based on VLAN.
+This functionality is Multus' trademark - adding additional interfaces to Pods.
+
+Following enhancement explores providing Multus CNI as an optional component to MicroShift.
+
+## Motivation
+
+Providing Multus CNI for MicroShift will help users:
+- wanting to integrate MicroShift into existing environments
+- wanting to slowly, step by step, migrate to MicroShift without being required to refactor everything at once
+- wanting to add additional interfaces because of newly discovered requirement
+
+### User Stories
+
+* As an MicroShift admin, I want to add additional interface to certain Pods so that I can
+  make them accessible over network that should not be available to rest of the cluster.
+* As an MicroShift admin, I want to add additional interface to certain Pods so that I can
+  access them directly without using Kubernetes' networking such as NodePorts, Load Balancers, Ingresses, etc.
+* As an MicroShift admin, I want to add additional interface to certain Pods so that I can
+  start slowly migrating existing solution to MicroShift.
+
+### Goals
+
+- Provide optional Multus CNI for MicroShift clusters that can be added to existing clusters
+
+### Non-Goals
+
+- Automatically removing Multus from the cluster upon RPM uninstall
+
+## Proposal
+
+Deliver Multus for MicroShift as an optional RPM containing required manifests that will be applied
+during MicroShift's start. There should be little to none changes to MicroShift itself as we want
+Multus CNI for MicroShift to fit optional components pattern.
+
+### Workflow Description
+
+**User** is a human user responsible for setting up and managing Edge Devices.
+**Application** is user's workload that intends to use additional interfaces.
+
+1. User gathers all information about the networking environment of the edge device.
+1. User prepares NetworkAttachmentDefinition (NAD) CRs that will allow Application to be part of the specified network.
+1. User prepares ostree commit that contains:
+   - (optional) Init procedures to configure OS for usage of the additional network
+   - MicroShift RPMs
+   - Multus for MicroShift RPM
+   - Application with the NetworkAttachmentDefinition CRs.
```

Comment:
These are just contents of the ostree commit - listing what's expected to make use of the Multus. Why should we clarify order here?

Reply:
The file locations and names matter, and this is the step where you're talking about adding the application to the ostree commit.

If you'd rather leave it out here, it should be explained somewhere later in the doc.

Reply:
I'm not sure I understand. You mentioned that we "need to make the order issue clear up here" - do you mean runtime order / scheduling components pre-MicroShift at the right time?

Reply:
Sorry, I wasn't clear.

The user has to carefully name their application manifests so they are loaded after Multus, if only so the NetworkAttachDefinition CRD is present at the point the application manifest that tries to use that API is loaded. Otherwise the loop to load manifests will block and startup will fail. That is not unique to Multus, of course, but we want to make sure that constraint is clearly spelled out in this design document, and eventually in the user docs.